### PR TITLE
do not write when getting default values

### PR DESCRIPTION
**Pull request by bimac**

### CODE REVIEW

**Comment by bimac** [comment]
@FlorianUekermann opinion?

**Comment by FlorianUekermann** [comment]
In principle I like getOrDefault. Useful pattern. Doing an implicit placement new is potentially questionable, but I need a refresher how EEPstore is used to be certain.

**Comment by bimac** [comment]
I use EEPstore as a way of storing parameters to EEPROM that ensures data integrity and, more importantly, compatibility. For instance, EEPstore falls back to default values when the EEPROM is unitialized or when the structure of the stored data has been changed across versions. I have used the pattern in several projects and have thus decided to package it as a library.

I typically use EEPstore to store data that is organized in a struct like so:

``` C++
typedef struct{
  uint8_t a = 42;
  uint8_t b = 23;
}storageVars;

storageVars p;
```

Retrieval from EEPROM - or default values:
``` C++
EEPstore<storageVars>::getOrDefault(p);
```

Writing to EEPROM:
``` C++
EEPstore<storageVars>::set(p);
```

Two questions came to mind:

1. Should there be an implicit write in `getOrDefault()` or should this only happen when explicitly calling `set()`?

2. How would you restore default values on demand ("restore factory settings")?

Ideally, I'd like to keep the library as flexible (and small) as possible ...

File: src/EEPstore.h [hunk 1]
```diff
@@ -49,18 +49,22 @@ template <class T> class EEPstore {
...
     }
   }
 
   static void set(const T &dataRef, const uint16_t address = 0) {
     EEPstore<T> storage(dataRef);
     EEPROM.put(address, storage);
   }
+
+  static inline __attribute__((always_inline)) void defaults(T &dataRef) {
+    new (dataRef) T;
```
Comment (bimac) [inline src/EEPstore.h]: This, I guess, will only work with typedef'd structs?

File: src/EEPstore.h [hunk 2]
```diff
@@ -49,22 +49,28 @@ template <class T> class EEPstore {
...
     EEPstore<T> storage(dataRef);
     EEPROM.get(address, storage);
-    if (storage.crc == storage.calcCRC()) {
+    bool valid = storage.crc == storage.calcCRC();
+    if (valid) {
       dataRef = storage.data;
-    } else {
+    }
+    return valid;
+  }
```
Comment (bimac) [inline src/EEPstore.h]: This would be an alternative to `getOrDefault()` that would not implicitly write to EEPROM

**Comment by bimac** [comment]
as discussed off-channel: I separated get() into getIfValid() and getOrSet(). The placement new was dropped.

**Pull request merged**